src/battle_util.c

```diff
@@ -4969,8 +4969,19 @@ u32 AbilityBattleEffects(u32 caseID, u32 battler, u32 ability, u32 special, u32
             gBattlerAttacker = battler;
             switch (gLastUsedAbility)
             {
+            case ABILITY_PICKUP:
+                if (gBattleMons[battler].item == ITEM_NONE
+                 && gBattleStruct->changedItems[battler] == ITEM_NONE   // Will not inherit an item
+                 && PickupHasValidTarget(battler))
+                {
+                    gBattlerTarget = RandomUniformExcept(RNG_PICKUP, 0, gBattlersCount, CantPickupItem);
```

Comment:
You are assigning a battler to `gBattlerTarget` but `tryrecycleitem` uses `gBattlerAttacker`. Does this always work?

Reply:
RandomUniform and RandomUniformExcept generate a number from arg2-arg3, so 0-gBattlersCount. Used gBattlersCount to make sure not to generate battlers that don't exist (in single battles)

Reply:
I think `RandomUniform` is inclusive, so you might want to pass `gBattlersCount - 1`? I'm not 100% sure, but a test that uses `PASSES_RANDOMLY` ought to be able to verify this.

Reply:
It says `RandomUniformExcept` is also inclusive but the test fails if using `gBattlersCount - 1`

![testfailed](https://github.com/user-attachments/assets/d088e809-a69c-4134-9b9e-e7b1593de255)

I think there is a bug in `RandomUniformExcept` since it only considered 2 battlers as possibilities but should consider 3

Reply:
Ahhh perhaps that's here: https://github.com/rh-hideout/pokeemerald-expansion/blob/master/test/test_runner_battle.c#L413? Should probably say `i <= hi`?